File: Backend/app/middleware/security.py
"""
🔒 Security Middleware for LinguaLink AI Backend
Rate limiting, input validation, and security headers
"""

import time
import hashlib
import logging
from typing import Dict, Optional
from collections import defaultdict, deque
from fastapi import Request, HTTPException, status
from fastapi.responses import JSONResponse
from starlette.middleware.base import BaseHTTPMiddleware
from starlette.responses import Response

from app.core.config import get_settings

logger = logging.getLogger(__name__)

class RateLimitMiddleware(BaseHTTPMiddleware):
    """Rate limiting middleware using sliding window algorithm"""
    
    def __init__(self, app, calls_per_minute: int = 60):
        super().__init__(app)
        self.calls_per_minute = calls_per_minute
        self.window_size = 60  # 1 minute window
        self.client_requests: Dict[str, deque] = defaultdict(deque)
        
    def _get_client_id(self, request: Request) -> str:
        """Get client identifier from request"""
        # Try to get real IP from headers (for reverse proxy setups)
        forwarded_for = request.headers.get("X-Forwarded-For")
        if forwarded_for:
            client_ip = forwarded_for.split(",")[0].strip()
        else:
            client_ip = request.client.host if request.client else "unknown"
        
        # Include user agent for better identification
        user_agent = request.headers.get("User-Agent", "")
        client_id = hashlib.md5(f"{client_ip}:{user_agent}".encode()).hexdigest()
        
        return client_id
    
    def _is_rate_limited(self, client_id: str) -> bool:
        """Check if client is rate limited using sliding window"""
        now = time.time()
        window_start = now - self.window_size
        
        # Get client's request history
        requests = self.client_requests[client_id]
        
        # Remove old requests outside the window
        while requests and requests[0] < window_start:
            requests.popleft()
        
        # Check if client has exceeded the limit
        if len(requests) >= self.calls_per_minute:
            return True
        
        # Add current request
        requests.append(now)
        
        return False
    
    async def dispatch(self, request: Request, call_next):
        """Process request with rate limiting"""
        settings = get_settings()
        
        # Skip rate limiting if disabled
        if not settings.ENABLE_RATE_LIMITING:
            return await call_next(request)
        
        # Skip rate limiting for health checks
        if request.url.path in ["/health", "/", "/docs", "/redoc"]:
            return await call_next(request)
        
        client_id = self._get_client_id(request)
        
        if self._is_rate_limited(client_id):
            logger.warning(f"Rate limit exceeded for client: {client_id[:8]}...")
            
            return JSONResponse(
                status_code=status.HTTP_429_TOO_MANY_REQUESTS,
                content={
                    "error": "Rate limit exceeded",
                    "message": f"Maximum {self.calls_per_minute} requests per minute allowed",
                    "retry_after": 60
                },
                headers={"Retry-After": "60"}
            )
        
        return await call_next(request)

class SecurityHeadersMiddleware(BaseHTTPMiddleware):
    """Add security headers to all responses"""
    
    async def dispatch(self, request: Request, call_next):
        """Add security headers to response"""
        response = await call_next(request)
        
        # Security headers
        security_headers = {
            "X-Content-Type-Options": "nosniff",
            "X-Frame-Options": "DENY",
            "X-XSS-Protection": "1; mode=block",
            "Referrer-Policy": "strict-origin-when-cross-origin",
            "Content-Security-Policy": "default-src 'self'",
            "Permissions-Policy": "geolocation=(), microphone=(), camera=()",
        }
        
        # Add headers to response
        for header, value in security_headers.items():
            response.headers[header] = value
        
        return response

class InputValidationMiddleware(BaseHTTPMiddleware):
    """Validate and sanitize input data"""
    
    def __init__(self, app):
        super().__init__(app)
        self.max_content_length = 10 * 1024 * 1024  # 10MB max request size
    
    async def dispatch(self, request: Request, call_next):
        """Validate request input"""
        
        # Check content length
        content_length = request.headers.get("content-length")
        if content_length and int(content_length) > self.max_content_length:
            logger.warning(f"Request too large: {content_length} bytes")
            return JSONResponse(
                status_code=status.HTTP_413_REQUEST_ENTITY_TOO_LARGE,
                content={
                    "error": "Request too large",
                    "message": f"Maximum request size is {self.max_content_length} bytes"
                }
            )
        
        # Validate content type for POST requests
        if request.method == "POST":
            content_type = request.headers.get("content-type", "")
            if not content_type.startswith("application/json"):
                logger.warning(f"Invalid content type: {content_type}")
                return JSONResponse(
                    status_code=status.HTTP_415_UNSUPPORTED_MEDIA_TYPE,
                    content={
                        "error": "Unsupported media type",
                        "message": "Content-Type must be application/json"
                    }
                )
        
        return await call_next(request)

class APIKeyMiddleware(BaseHTTPMiddleware):
    """API key authentication middleware"""
    
    def __init__(self, app):
        super().__init__(app)
        self.settings = get_settings()
    
    async def dispatch(self, request: Request, call_next):
        """Validate API key if authentication is enabled"""
        
        # Skip API key validation if disabled
        if not self.settings.ENABLE_API_KEY_AUTH:
            return await call_next(request)
        
        # Skip API key for public endpoints
        public_endpoints = ["/", "/health", "/docs", "/redoc", "/openapi.json"]
        if request.url.path in public_endpoints:
            return await call_next(request)
        
        # Check for API key
        api_key = request.headers.get(self.settings.API_KEY_HEADER)
        
        if not api_key:
            logger.warning("Missing API key")
            return JSONResponse(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content={
                    "error": "Authentication required",
                    "message": f"API key required in {self.settings.API_KEY_HEADER} header"
                }
            )
        
        # Validate API key
        if self.settings.ALLOWED_API_KEYS and api_key not in self.settings.ALLOWED_API_KEYS:
            logger.warning(f"Invalid API key: {api_key[:8]}...")
            return JSONResponse(
                status_code=status.HTTP_401_UNAUTHORIZED,
                content={
                    "error": "Invalid API key",
                    "message": "The provided API key is not valid"
                }
            )
        
        return await call_next(request)

class LoggingMiddleware(BaseHTTPMiddleware):
    """Enhanced logging middleware for production"""
    
    async def dispatch(self, request: Request, call_next):
        """Log request and response details"""
        start_time = time.time()
        
        # Log request
        client_ip = request.headers.get("X-Forwarded-For", request.client.host if request.client else "unknown")
        logger.info(f"📥 {request.method} {request.url.path} from {client_ip}")
        
        # Process request
        try:
            response = await call_next(request)
            
            # Log response
            process_time = time.time() - start_time
            logger.info(f"📤 {response.status_code} {request.url.path} ({process_time:.3f}s)")
            
            # Add processing time header
            response.headers["X-Process-Time"] = str(process_time)
            
            return response
            
        except Exception as e:
            process_time = time.time() - start_time
            logger.error(f"❌ {request.method} {request.url.path} failed after {process_time:.3f}s: {e}")
            raise

File: src/middleware/security.ts
/**
 * 🔒 ENTERPRISE SECURITY MIDDLEWARE
 * Implements comprehensive security headers and protection mechanisms
 * Following OWASP security best practices
 */

import { NextRequest, NextResponse } from 'next/server';
import { secureLogger, SecurityEventType, SecuritySeverity } from '@/lib/secure-logger';

// 🛡️ SECURITY HEADERS CONFIGURATION
const SECURITY_HEADERS = {
  // Content Security Policy - Prevent XSS attacks
  'Content-Security-Policy': [
    "default-src 'self'",
    "script-src 'self' 'unsafe-inline' 'unsafe-eval' https://clerk.com https://*.clerk.com https://challenges.cloudflare.com",
    "style-src 'self' 'unsafe-inline' https://fonts.googleapis.com",
    "font-src 'self' https://fonts.gstatic.com",
    "img-src 'self' data: blob: https: http:",
    "media-src 'self' blob: data:",
    "connect-src 'self' https://clerk.com https://*.clerk.com wss: ws: https://api.openai.com https://api.featherless.ai",
    "frame-src 'self' https://challenges.cloudflare.com",
    "object-src 'none'",
    "base-uri 'self'",
    "form-action 'self'",
    "frame-ancestors 'none'",
    "upgrade-insecure-requests"
  ].join('; '),

  // Prevent clickjacking attacks
  'X-Frame-Options': 'DENY',

  // Prevent MIME type sniffing
  'X-Content-Type-Options': 'nosniff',

  // Enable XSS protection
  'X-XSS-Protection': '1; mode=block',

  // Referrer policy
  'Referrer-Policy': 'strict-origin-when-cross-origin',

  // Permissions policy
  'Permissions-Policy': [
    'camera=()',
    'microphone=()',
    'geolocation=()',
    'payment=()',
    'usb=()',
    'magnetometer=()',
    'gyroscope=()',
    'accelerometer=()'
  ].join(', '),

  // Strict Transport Security (HTTPS only)
  'Strict-Transport-Security': 'max-age=31536000; includeSubDomains; preload',

  // Cross-Origin policies
  'Cross-Origin-Embedder-Policy': 'require-corp',
  'Cross-Origin-Opener-Policy': 'same-origin',
  'Cross-Origin-Resource-Policy': 'same-origin',

  // Cache control for sensitive pages
  'Cache-Control': 'no-store, no-cache, must-revalidate, proxy-revalidate',
  'Pragma': 'no-cache',
  'Expires': '0',

  // Remove server information
  'Server': 'LinguaLink',

  // Custom security headers
  'X-Powered-By': 'LinguaLink Security Framework',
  'X-Security-Level': 'Enterprise'
};

// 🚨 RATE LIMITING CONFIGURATION
const RATE_LIMITS = {
  // API endpoints
  '/api/': { requests: 100, window: 60000 }, // 100 requests per minute
  '/api/messages': { requests: 50, window: 60000 }, // 50 messages per minute
  '/api/auth/': { requests: 10, window: 60000 }, // 10 auth requests per minute
  '/api/users/search': { requests: 20, window: 60000 }, // 20 searches per minute
  
  // File uploads
  '/api/upload': { requests: 10, window: 300000 }, // 10 uploads per 5 minutes
  
  // Default for all other routes
  'default': { requests: 200, window: 60000 } // 200 requests per minute
};

// 🔍 SUSPICIOUS PATTERNS TO DETECT (REFINED FOR PRODUCTION)
const SUSPICIOUS_PATTERNS = [
  // SQL Injection attempts (more specific)
  /(\bunion\s+select\b)/i,
  /(\bselect\s+.*\bfrom\s+)/i,
  /(\bdrop\s+table\b)/i,
  /(\bexec\s*\()/i,
  /(\b(or|and)\s+\d+\s*=\s*\d+\s*(--|\#))/i,

  // XSS attempts (more specific)
  /<script[^>]*>.*?<\/script>/gi,
  /javascript\s*:/i,
  /on(load|error|click|mouseover)\s*=/i,

  // Path traversal (more specific)
  /\.\.\/(.*\/){3,}/,
  /\.\.\\\.\.\\/,

  // Command injection (more specific)
  /;\s*(rm|del|format|shutdown)/i,
  /\|\s*(nc|netcat|wget|curl)\s/i,

  // NoSQL injection (more specific)
  /\$where\s*:/i,
  /\{\s*\$ne\s*:/i
];

// 🗂️ RATE LIMITING STORE
const rateLimitStore = new Map<string, { count: number; resetTime: number }>();

/**
 * 🔒 MAIN SECURITY MIDDLEWARE
 */
export function securityMiddleware(request: NextRequest): NextResponse {
  const response = NextResponse.next();
  const clientIP = getClientIP(request);
  const userAgent = request.headers.get('user-agent') || '';
  const pathname = request.nextUrl.pathname;
  const method = request.method;

  try {
    // 🛡️ APPLY SECURITY HEADERS
    Object.entries(SECURITY_HEADERS).forEach(([key, value]) => {
      response.headers.set(key, value);
    });

    // 🚨 DETECT SUSPICIOUS ACTIVITY
    const suspiciousActivity = detectSuspiciousActivity(request);
    if (suspiciousActivity.length > 0) {
      secureLogger.logCriticalSecurity(
        SecurityEventType.SUSPICIOUS_ACTIVITY,
        'Suspicious activity detected and blocked',
        {
          clientIP: clientIP?.substring(0, 8) + '***',
          userAgent: userAgent.substring(0, 50) + '...',
          pathname,
          method,
          patterns: suspiciousActivity,
          action: 'BLOCKED'
        }
      );

      // Block the request
      return new NextResponse('Access Denied', { 
        status: 403,
        headers: {
          'Content-Type': 'text/plain',
          ...Object.fromEntries(Object.entries(SECURITY_HEADERS))
        }
      });
    }

    // 🚦 RATE LIMITING
    const rateLimitResult = checkRateLimit(clientIP, pathname);
    if (!rateLimitResult.allowed) {
      secureLogger.logWarningSecurity(
        SecurityEventType.RATE_LIMIT_EXCEEDED,
        'Rate limit exceeded',
        {
          clientIP: clientIP?.substring(0, 8) + '***',
          pathname,
          method,
          limit: rateLimitResult.limit,
          window: rateLimitResult.window,
          action: 'RATE_LIMITED'
        }
      );

      return new NextResponse('Rate Limit Exceeded', { 
        status: 429,
        headers: {
          'Content-Type': 'text/plain',
          'Retry-After': Math.ceil(rateLimitResult.retryAfter / 1000).toString(),
          ...Object.fromEntries(Object.entries(SECURITY_HEADERS))
        }
      });
    }

    // 🔍 LOG SECURITY EVENT
    secureLogger.logInfoSecurity(
      SecurityEventType.DATA_ACCESS,
      'Request processed successfully',
      {
        clientIP: clientIP?.substring(0, 8) + '***',
        pathname,
        method,
        userAgent: userAgent.substring(0, 50) + '...',
        action: 'ALLOWED'
      }
    );

    return response;

  } catch (error) {
    secureLogger.logCriticalSecurity(
      SecurityEventType.SYSTEM_ERROR,
      'Security middleware error',
      {
        clientIP: clientIP?.substring(0, 8) + '***',
        pathname,
        method,
        error: 'MIDDLEWARE_ERROR'
      }
    );

    // Fail securely - apply security headers even on error
    Object.entries(SECURITY_HEADERS).forEach(([key, value]) => {
      response.headers.set(key, value);
    });

    return response;
  }
}

/**
 * 🌐 GET CLIENT IP ADDRESS
 */
function getClientIP(request: NextRequest): string | null {
  // Check various headers for the real IP
  const forwardedFor = request.headers.get('x-forwarded-for');
  const realIP = request.headers.get('x-real-ip');
  const cfConnectingIP = request.headers.get('cf-connecting-ip');
  
  if (cfConnectingIP) return cfConnectingIP;
  if (realIP) return realIP;
  if (forwardedFor) return forwardedFor.split(',')[0].trim();
  
  // NextRequest doesn't have ip property, return null as fallback
  return null;
}

/**
 * 🕵️ DETECT SUSPICIOUS ACTIVITY (REFINED FOR PRODUCTION)
 */
function detectSuspiciousActivity(request: NextRequest): string[] {
  const suspiciousPatterns: string[] = [];
  const pathname = request.nextUrl.pathname;
  const userAgent = request.headers.get('user-agent') || '';

  // 🚫 SKIP SECURITY CHECKS FOR LEGITIMATE REQUESTS
  const isLegitimateRequest =
    pathname.startsWith('/_next/') ||
    pathname.startsWith('/static/') ||
    pathname === '/sw.js' ||
    pathname === '/dashboard' ||
    pathname === '/favicon.ico' ||
    pathname.includes('.well-known') ||
    (userAgent.includes('Mozilla') && userAgent.includes('Chrome')) ||
    (userAgent.includes('Mozilla') && userAgent.includes('Safari')) ||
    (userAgent.includes('Mozilla') && userAgent.includes('Firefox'));

  if (isLegitimateRequest) {
    return []; // Skip detection for legitimate browser requests
  }

  // 🔍 CHECK URL FOR ACTUAL THREATS ONLY
  const url = request.url;
  SUSPICIOUS_PATTERNS.forEach((pattern, index) => {
    if (pattern.test(url)) {
      suspiciousPatterns.push(`URL_THREAT_${index + 1}`);
    }
  });

  // 🔍 CHECK ONLY SPECIFIC HEADERS THAT COULD CONTAIN THREATS
  const criticalHeaders = ['authorization', 'cookie', 'x-forwarded-for', 'x-real-ip'];
  criticalHeaders.forEach(headerName => {
    const headerValue = request.headers.get(headerName);
    if (headerValue) {
      SUSPICIOUS_PATTERNS.forEach((pattern, index) => {
        if (pattern.test(headerValue)) {
          suspiciousPatterns.push(`HEADER_${headerName.toUpperCase()}_THREAT_${index + 1}`);
        }
      });
    }
  });

  return suspiciousPatterns;
}

/**
 * 🚦 CHECK RATE LIMITING
 */
function checkRateLimit(clientIP: string | null, pathname: string): {
  allowed: boolean;
  limit: number;
  window: number;
  retryAfter: number;
} {
  if (!clientIP) {
    return { allowed: true, limit: 0, window: 0, retryAfter: 0 };
  }

  // Find matching rate limit rule
  let rateLimit = RATE_LIMITS.default;
  for (const [path, limit] of Object.entries(RATE_LIMITS)) {
    if (path !== 'default' && pathname.startsWith(path)) {
      rateLimit = limit;
      break;
    }
  }

  const key = `${clientIP}:${pathname}`;
  const now = Date.now();
  const windowStart = now - rateLimit.window;

  // Clean up old entries
  if (rateLimitStore.size > 10000) {
    for (const [k, v] of rateLimitStore.entries()) {
      if (v.resetTime < now) {
        rateLimitStore.delete(k);
      }
    }
  }

  const current = rateLimitStore.get(key);
  
  if (!current || current.resetTime < now) {
    // First request or window expired
    rateLimitStore.set(key, {
      count: 1,
      resetTime: now + rateLimit.window
    });
    return { allowed: true, limit: rateLimit.requests, window: rateLimit.window, retryAfter: 0 };
  }

  if (current.count >= rateLimit.requests) {
    // Rate limit exceeded
    return {
      allowed: false,
      limit: rateLimit.requests,
      window: rateLimit.window,
      retryAfter: current.resetTime - now
    };
  }

  // Increment counter
  current.count++;
  return { allowed: true, limit: rateLimit.requests, window: rateLimit.window, retryAfter: 0 };
}

/**
 * 🧹 SANITIZE INPUT DATA
 */
export function sanitizeInput(input: any): any {
  if (typeof input === 'string') {
    return input
      .replace(/<script\b[^<]*(?:(?!<\/script>)<[^<]*)*<\/script>/gi, '')
      .replace(/javascript:/gi, '')
      .replace(/on\w+\s*=/gi, '')
      .replace(/[<>]/g, '')
      .trim()
      .substring(0, 10000); // Limit length
  }
  
  if (Array.isArray(input)) {
    return input.map(sanitizeInput);
  }
  
  if (typeof input === 'object' && input !== null) {
    const sanitized: any = {};
    for (const [key, value] of Object.entries(input)) {
      sanitized[sanitizeInput(key)] = sanitizeInput(value);
    }
    return sanitized;
  }
  
  return input;
}

/**
 * 🔍 VALIDATE INPUT AGAINST SCHEMA
 */
export function validateInput(input: any, schema: any): { valid: boolean; errors: string[] } {
  const errors: string[] = [];
  
  // Basic validation - extend as needed
  if (schema.required && !input) {
    errors.push('Required field is missing');
  }
  
  if (schema.maxLength && typeof input === 'string' && input.length > schema.maxLength) {
    errors.push(`Input exceeds maximum length of ${schema.maxLength}`);
  }
  
  if (schema.pattern && typeof input === 'string' && !schema.pattern.test(input)) {
    errors.push('Input does not match required pattern');
  }
  
  return { valid: errors.length === 0, errors };
}
